embedchain/vectordb/deeplake.py

```diff
@@ -0,0 +1,192 @@
+
+    def add(
+        self,
+        embeddings: List[List[float]],
+        documents: List[str],
+        metadatas: List[object],
+        ids: List[str],
+        skip_embedding: bool,
+    ):
+        """add data in vector database
+        :param embeddings: list of embeddings for the corresponding documents to be added
+        :type documents: List[List[float]]
+        :param documents: list of texts to add
+        :type documents: List[str]
+        :param metadatas: list of metadata associated with docs
+        :type metadatas: List[object]
+        :param ids: ids of docs
+        :type ids: List[str]
+        :param skip_embedding: A boolean flag indicating if the embedding for the documents to be added is to be
+        generated or not
+        :type skip_embedding: bool
+        """
+        if not skip_embedding:
+            embeddings = self.embedder.embedding_fn(documents)
+
+        payloads = []
+        for id, document, metadata in zip(ids, documents, metadatas):
+            metadata["identifier"] = id
+            payloads.append(copy.deepcopy(metadata))
+
+        for i in range(0, len(metadatas), self.BATCH_SIZE):
+            self.client.add(
+                text=documents[i : i + self.BATCH_SIZE],
+                embedding=embeddings[i : i + self.BATCH_SIZE],
+                metadata=payloads[i : i + self.BATCH_SIZE],
+            )
+
+    def query(self, input_query: List[str], n_results: int, where: Dict[str, any], skip_embedding: bool) -> List[str]:
+        """
+        query contents from vector database based on vector similarity
+        :param input_query: list of query string
+        :type input_query: List[str]
+        :param n_results: no of similar documents to fetch from database
+        :type n_results: int
+        :param where: Optional. to filter data
+        :type where: Dict[str, any]
+        :param skip_embedding: A boolean flag indicating if the embedding for the documents to be added is to be
+        generated or not
+        :type skip_embedding: bool
+        :return: Database contents that are the result of the query
+        :rtype: List[str]
+        """
+        if not skip_embedding:
+            query_vector = self.embedder.embedding_fn([input_query])[0]
+        else:
+            query_vector = input_query
+
+        keys = set(where.keys() if where is not None else set())
+
+        deeplake_metadata_filter = {}
+        if len(keys.intersection(self.metadata_keys)) != 0:
+            for key in keys.intersection(self.metadata_keys):
+                deeplake_metadata_filter[key] = where.get(key)
+
+        results = self.client.search(
+            embedding=query_vector,
+            exec_option=self.config.exec_option,
+            filter={"metadata": deeplake_metadata_filter},
+            k=n_results,
+        )
+        response = []
+        for result in results.get("text", []):
+            response.append(result)
+        return response
```

Comment:
@rupeshbansal The query function should now return list of tuples(context, source_url, doc_id) instead of return just the list of context. Check [this](https://github.com/embedchain/embedchain/pull/831). Can you update the function appropriately?